app/controllers/application_controller.rb

```diff
@@ -9,40 +9,28 @@ class ApplicationController < ActionController::Base
   # optionally have authenticated access to public routes.
   # for example, /auctions/:id/bids will unveil bidder info about
   # the authenticated user. but the page also works fine sans authentication.
-  before_action :set_current_user_to_api_user!, if: proc { api_request? }
-
-  def current_user
-    @current_user ||= User.where(id: session[:user_id]).first
-  end
-
-  def require_authentication
-    if html_request?
-      redirect_if_not_logged_in!
-    elsif api_request?
-      set_current_user_to_api_user!(raise_errors: true)
-    end
-  end
+  before_action :set_current_user, if: proc { api_request? }
 
   def require_admin
-    if html_request?
-      # if the request is via the web UI,
-      # we need 'and return' to ensure the redirect happens.
-      require_authentication and return
-    elsif api_request?
-      # otherwise, if it's an API request, we don't need to redirect.
-      require_authentication
-    end
+    require_authentication
     is_admin = Admins.verify?(github_id)
     fail UnauthorizedError::MustBeAdmin unless is_admin
 
     is_admin
   end
 
+  delegate :require_authentication, :current_user, :github_id, :set_current_user,
+           to: :authenticator
+  
   rescue_from 'UnauthorizedError::MustBeAdmin' do |error|
     message = error.message || "Unauthorized"
     handle_error(message)
   end
 
+  rescue_from 'UnauthorizedError::RedirectToLogin' do |error|
```

Comment:
This is a case where we have an Authenticator raise an exception to have the controller take an action